demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c

```diff
@@ -0,0 +1,613 @@
+                                     sizeof( keyCLIENT_CERTIFICATE_PEM ),
+                                     ( uint8_t * ) pkcs11configLABEL_DEVICE_CERTIFICATE_FOR_TLS,
+                                     &xObject );
+
+    if( ( xResult != CKR_OK ) || ( xObject == CK_INVALID_HANDLE ) )
+    {
+        configPRINTF( ( "ERROR: Failed to provision device certificate. %d \r\n", xResult ) );
+    }
+
+    /* Provisioning JITR CA Certificate */
+    if( xResult == CKR_OK )
+    {
+        if( sizeof( keyJITR_DEVICE_CERTIFICATE_AUTHORITY_PEM ) != 0 )
+        {
+            xResult = xProvisionCertificate( xGlobalSession,
+                                             keyJITR_DEVICE_CERTIFICATE_AUTHORITY_PEM,
+                                             sizeof( keyJITR_DEVICE_CERTIFICATE_AUTHORITY_PEM ),
+                                             ( uint8_t * ) pkcs11configLABEL_JITP_CERTIFICATE,
+                                             &xObject );
+
+            xResult = CKR_OK;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        configPRINTF( ( "Device credential provisioning succeeded.\r\n" ) );
+    }
+    else
+    {
+        configPRINTF( ( "Device credential provisioning failed.\r\n" ) );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+/* @brief Provisions a Device Using Internally Generated Keys
+ *
+ *
+ *  \return Void function, console outputs upon success/failure
+ *
+ */
+static void vDevModeDeviceKeyProvisioning( void )
+{
+    xResult = xInitializePkcs11Session( &xGlobalSession );
+
+    if( xResult != CKR_OK )
+    {
+        configPRINTF( ( "Failed to open PKCS #11 session.\r\n" ) );
+    }
+
+    xResult = C_GetFunctionList( &pxGlobalFunctionList );
+
+    if( xResult != CKR_OK )
+    {
+        configPRINTF( ( "Failed to get function list.\r\n" ) );
+    }
+
+    #if ( !DEMO_PART )
+        {
+            CredentialsProvisioned_t xCurrentCredentials = eStateUnknown;
+
+            /* Note that mbedTLS does not permit a signature on all 0's. */
+            CK_BYTE xHashedMessage[ SHA256_DIGEST_SIZE ] = { 0xab };
+            CK_MECHANISM xMechanism;
+            CK_BYTE xSignature[ RSA_SIGNATURE_SIZE ] = { 0 };
+            CK_BYTE xEcPoint[ 256 ] = { 0 };
+            CK_BYTE xEcParams[ 11 ] = { 0 };
+            CK_KEY_TYPE xKeyType;
+            CK_ULONG xSignatureLength;
+            CK_ATTRIBUTE xTemplate;
+            CK_OBJECT_CLASS xClass;
+
+            /* mbedTLS structures for verification. */
+            int lMbedTLSResult;
+            int ret;
+            mbedtls_ecdsa_context xEcdsaContext;
+            uint8_t ucSecp256r1Oid[] = pkcs11DER_ENCODED_OID_P256; /*"\x06\x08" MBEDTLS_OID_EC_GRP_SECP256R1; */
+
+            xResult = xDestroyCredentials( xGlobalSession );
+
+            if( xResult != CKR_OK )
+            {
+                configPRINTF( ( "Failed to destroy credentials before Generating Key Pair.\r\n" ) );
```

Comment:
For this (and other configPRINTFs in this file) consider printing the error code returned to help with debugging using logs.

For example, configPRINTF( ( "ERROR: %d - Failed to destroy credentials before Generating Key Pair.\r\n", xResult ) );

Reply:
Modified all my config outputs to include error codes! Thanks.